File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System;
using System.Collections.Generic;

class Solution {
    public static void bfs(int startX,int startY,int[,]map)
    {
        
        int[] dx = new int[4] { -1, 1, 0, 0 };
        int[] dy = new int[4] { 0, 0, -1, 1 };
        Queue<Tuple<int, int>> q = new Queue<Tuple<int, int>>();
             
        q.Enqueue(new Tuple<int, int>(startX, startY));
        while(q.Count!=0)
        {
             Tuple<int, int> cur = q.Dequeue();
             //int nowX = cur.Item1;
             //int nowY = cur.Item2;
            for(int i=0;i<4;i++)
            {
                  int nextX = cur.Item1 + dx[i];
                  int nextY = cur.Item2 + dy[i];
               
                
             if (0 <= nextX && nextX < map.GetLength(0)&& 0 <= nextY && nextY < map.GetLength(1) && map[nextX, nextY] == 1)
            {
             q.Enqueue(new Tuple<int, int>(nextX, nextY));
                 map[nextX,nextY]=map[cur.Item1,cur.Item2]+1;
            }
                
                
        }
        }
    }
    public int solution(int[,] maps) {
        int a=maps.GetLength(0);
        int b=maps.GetLength(1);
   
        bfs(0,0,maps);
        
        if(maps[a-1,b-1]==1)
            return -1;
        else
            return (maps[a-1,b-1]);
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
using namespace std;

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    int dx[4]={-1,0,1,0};
    int dy[4]={0,-1,0,1};
    int n = maps.size();
    int m = maps[0].size();
    queue<pair<int,int>> q;
    vector<vector<int>> dist(n,vector<int>(m,-1));
    q.push({0,0});
    dist[0][0]=1;
    while(!q.empty())
    {
        pair<int,int> cur= q.front();
        q.pop();
        int cx=cur.first;
        int cy=cur.second;
        for(int i=0;i<4;i++)
        {
            int nx = cx+dx[i];
            int ny = cy+dy[i];
            if(nx<0||nx>=n||ny<0||ny>=m) continue;
            if(maps[nx][ny]==0||dist[nx][ny]!=-1) continue;
            dist[nx][ny]=dist[cx][cy]+1;
            q.push({nx,ny});
        }
    }
   answer=dist[n-1][m-1];
    
    return answer;
}